libp2p/protocols/pubsub/gossipsub/scoring.nim

```diff
@@ -85,27 +90,48 @@ proc colocationFactor(g: GossipSub, peer: PubSubPeer): float64 =
 
 {.pop.}
 
-proc disconnectPeer(g: GossipSub, peer: PubSubPeer) {.async.} =
-  let agent =
+proc getAgent(peer: PubSubPeer): string =
+  return
     when defined(libp2p_agents_metrics):
       if peer.shortAgent.len > 0:
         peer.shortAgent
       else:
         "unknown"
     else:
       "unknown"
+
+proc disconnectPeer(g: GossipSub, peer: PubSubPeer) {.async.} =
+  let agent = getAgent(peer)
   libp2p_gossipsub_bad_score_disconnection.inc(labelValues = [agent])
 
   try:
     await g.switch.disconnect(peer.peerId)
   except CatchableError as exc: # Never cancelled
     trace "Failed to close connection", peer, error = exc.name, msg = exc.msg
 
-proc disconnectBadPeerCheck*(g: GossipSub, peer: PubSubPeer, score: float64) =
-  if g.parameters.disconnectBadPeers and score < g.parameters.graylistThreshold and
+proc disconnectIfBadPeer*(g: GossipSub, peer: PubSubPeer, score: float64, threshold: float64): bool =
+  if g.parameters.disconnectBadPeers and score < threshold and
      peer.peerId notin g.parameters.directPeers:
-    debug "disconnecting bad score peer", peer, score = peer.score
+    debug "disconnecting bad score peer", peer, score, threshold
     asyncSpawn(g.disconnectPeer(peer))
+    return true
+
+  return false
+
+proc disconnectIfBadTrafficPeer*(g: GossipSub, peer: PubSubPeer) =
+  if peer.totalTraffic > 0:
+    # dividing in this way to avoid integer overflow
+    let agent = getAgent(peer)
+    let invalidTrafficRatio = float64(peer.invalidTraffic) / float64(peer.totalTraffic)
```

Comment:
We need to have these value expire at some point (or continuously reduced), otherwise I can be a good peer for a few days, and then use my accumulated score to spam you heavily in a short amount of time